### ISSUE TITLE
Angular Elements: Can't bind to custom element's properties.

### ISSUE THREAD

**Issue by dali-trabelsi** [angular/angular#48303]
### Which @angular/* package(s) are the source of the bug?

elements

### Is this a regression?

No

### Description

I created a custom element using Angular 14, and I am trying to use it inside another angular project that is using version 13.
I have a problem with data binding my custom element's properties to some variables. 

In my custom elements, I have declared `@Input() public previewMode: boolean; `
When I call my custom component without binding, it works fine (like this `<custom-element preview-mode="false" ><custom-element >`) 
The custom element correctly receives the value of previewMode, but when I try to bind it to a variable it doesn't work and its always undefined. I tried the following binding methods but to no avail:
`<custom-element [preview-mode]="booleanVar" ><custom-element>`
`<custom-element [previewMode]="booleanVar" ><custom-element>`
`<custom-element preview-mode="{{booleanVar}}" ><custom-element>`.
Am I missing something here because I don't seem to find any other way of doing this.

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Parent project: 

Angular CLI: 13.3.9
Node: 16.18.0
Package Manager: npm 8.19.2
OS: win32 x64

Angular: 13.3.11
... animations, common, compiler, compiler-cli, core, forms
... localize, platform-browser, platform-browser-dynamic, router
... service-worker

Package                         Version
------------------------------------------------------------
@angular-devkit/architect          0.1303.9
@angular-devkit/build-angular      13.3.9
@angular-devkit/core               13.3.9
@angular-devkit/schematics         13.3.9
@angular/cdk                       13.3.9
@angular/cli                       13.3.9
@angular/material                  13.3.9
@angular/material-moment-adapter   13.3.9
@schematics/angular                13.3.9
rxjs                               6.6.7
typescript                         4.6.4

-------------------------------------------------------------------------------------------

Web Component Project:

Angular CLI: 14.1.3
Node: 16.18.0
Package Manager: npm 8.19.2
OS: win32 x64

Angular: 14.1.2
... animations, cdk, common, compiler, compiler-cli, core
... elements, forms, platform-browser, platform-browser-dynamic
... router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1401.3
@angular-devkit/build-angular   14.1.3
@angular-devkit/core            14.1.3
@angular-devkit/schematics      14.1.3
@angular/cli                    14.1.3
@schematics/angular             14.1.3
rxjs                            7.5.6
typescript                      4.7.4
```

**Comment by pkozlowski-opensource** [comment]
@dali-trabelsi could you please share a minimal reproduce scenario?

**Comment by dali-trabelsi** [comment]
@pkozlowski-opensource  here is a codesandbox with minimal reproduction of the bug: [https://codesandbox.io/s/amazing-butterfly-dwdwi9?file=/src/app/app.component.html](https://codesandbox.io/s/amazing-butterfly-dwdwi9?file=/src/app/app.component.html)

For the webcomponent I used a freshly created one using Angular Elements 15, below is the component's code:

HTML: `<h1>Hello {{ title }}</h1>` 

TS File: 
```
export class AppComponent implements OnInit, AfterViewChecked, AfterViewInit {
  @Input() title?: string;

  ngOnInit(): void {
    alert('Hello ' + this.title + ' from ngOnInit');
  }

  ngAfterViewChecked(): void {
    alert('Hello ' + this.title + ' from ngAfterViewChecked');
  }

  ngAfterViewInit(): void {
    alert('Hello ' + this.title + ' from ngAfterViewInit');
  }
}
```

**Comment by dali-trabelsi** [comment]
@pkozlowski-opensource  this seems to be a problem with angular elements compiling itself, any idea please ?

**Comment by dali-trabelsi** [comment]
> @dali-trabelsi the reproduction provided, with a precompiled (and minified!) custom element is not a minimal reproduction. We should be able to see what's going on in the custom element.

Hi @alxhub, thanks for the reply. I already provided the full code of the web component in a comment to @pkozlowski-opensource below here: 

> @pkozlowski-opensource here is a codesandbox with minimal reproduction of the bug: https://codesandbox.io/s/amazing-butterfly-dwdwi9?file=/src/app/app.component.html
> 
> For the webcomponent I created a new one for the reproduction using Angular Elements 15, below is the component's entire code which was compiled and used as a web component in the reproduction link.
> 
> ```
> @Component({
>   template: `<h1>Hello {{ title }}</h1>` 
> })
> export class AppComponent implements OnInit, AfterViewChecked, AfterViewInit {
> 
>   @Input() title?: string;
> 
>   ngOnInit(): void {
>     alert('Hello ' + this.title + ' from ngOnInit');
>   }
> 
>   ngAfterViewChecked(): void {
>     alert('Hello ' + this.title + ' from ngAfterViewChecked');
>   }
> 
>   ngAfterViewInit(): void {
>     alert('Hello ' + this.title + ' from ngAfterViewInit');
>   }
> }
> ```

Is there anything else you recommend for giving a better minimal reproduction? Otherwise I hope its clear what's going on and why I think its a bug.

**Comment by alan-agius4** [comment]
Here is another reproduction from #49833 
https://github.com/HazzMan2409/ng-elements-input-issue

**Comment by MatGarreau** [comment]
Hello, is there an update on this issue ?

**Comment by ValerieBrave** [comment]
Hi, faced the same issue with Angular v15 
For me it was input's naming, it seems that properties with upper letters and underscores are not read properly

`formId` undefined
`form_id` undefined
`form` OK

Hope this helps

**Comment by typescripterpro** [comment]
Maybe this can help https://typescripter.space/questions/how-angular-services-transmit-information-to-components

**Comment by jeanpaulattard** [comment]
Are there any official stance on this? Is this something that is being considered? Given that it's been present for such a long time.

**Comment by JoostK** [comment]
This behavior occurs because Angular separates its element creation phase from the update phase. When the custom element has already been registered will the creation of the DOM element immediately connect to the custom element implementation, which proceeds by instantiating the Angular component instance _and_ triggers a round of change detection. That is the root cause here, as this means that the outer template hasn't had a chance to evaluate the element's attributes yet so none of the inputs are available.

The "fix" might be to remove the eager change detection during `connect`, but that would likely be a large breaking change and I'm not sure how that would behave; I suspect it will misbehave in other manners due to depending on an explicit tick of the `ApplicationRef` the custom element was registered using.

**Comment by engineer-fortuitytech** [comment]
> Hi, faced the same issue with Angular v15 For me it was input's naming, it seems that properties with upper letters and underscores are not read properly
> 
> `formId` undefined `form_id` undefined `form` OK
> 
> Hope this helps

this worked for me as well - renamed all of my inputs lowercase and all lifecycle hooks picked up the input values.

**Comment by Moustingle** [comment]
I had this problem on dynamically created components.
I managed to get the data inside the ngOnInit function by wrapping the logic inside a SetTimeout.

If I reproduce the first example :

```
@Component({
  template: '<h1>Hello {{ title }}</h1>'
})
export class AppComponent implements OnInit {

  @Input() title?: string;

  ngOnInit(): void {

    setTimeout(() => {
      console.log(this.title); // ---> 'title input data'
    }, 0);

    console.log(this.title); // ---> undefined
  }

}
```

If it is of any help.

**Comment by c-harding** [comment]
Another reproduction, with examples of ngOnInit, effects and resources: https://github.com/angular/angular/issues/66588

While a `setTimeout` in `ngOnInit` does work, more fundamental features such as `.required` are also broken. This limitation should at least be clearly documented in Angular Elements.

**Comment by c-harding** [comment]
@JoostK

> The "fix" might be to remove the eager change detection during connect, but that would likely be a large breaking change and I'm not sure how that would behave; I suspect it will misbehave in other manners due to depending on an explicit tick of the ApplicationRef the custom element was registered using.

Would it be possible to provide the first values _before_ instantiating the instance instead? The workaround I'm currently using involves calling `document.createElement`, `setAttribute`, and then adding the element to the DOM, which does work and sets the inputs before `connect` is called.

For me it would be acceptable if this only applied to `[attr.]` attributes, to avoid breaking logic of setting properties rather than attributes.

**Comment by JoostK** [comment]
> [@JoostK](https://github.com/JoostK)
> 
> > The "fix" might be to remove the eager change detection during connect, but that would likely be a large breaking change and I'm not sure how that would behave; I suspect it will misbehave in other manners due to depending on an explicit tick of the ApplicationRef the custom element was registered using.
> 
> Would it be possible to provide the first values _before_ instantiating the instance instead? The workaround I'm currently using involves calling `document.createElement`, `setAttribute`, and then adding the element to the DOM, which does work and sets the inputs before `connect` is called.
> 
> For me it would be acceptable if this only applied to `[attr.]` attributes, to avoid breaking logic of setting properties rather than attributes.

At least when Angular is used to create the DOM element on which a custom element is bootstrapped, that would be quite fundamentally against Angular's two-phase DOM rendering approach: creation mode followed by update mode, the latter of which will only execute during a CD cycle. Element creation and adding to the document all happen during the creation phase.

**Comment by c-harding** [comment]
Yeah, it's true that this does go against the typical Angular instantiation order. However, I'd argue that the current behaviour goes against it too, as ngOnInit is called without any of the dynamic inputs it's supposed to be called with.
